### DISCUSSION TITLE
What does the data in a collision event mean?

### DISCUSSION THREAD

**Discussion by a-type** [pmndrs/use-cannon#130] [Q&A] [answered]
The shape of the event provided for a collision is documented, but the meanings of properties are not:

```
type Event = {
  op: string
  type: string
  body: THREE.Object3D
  target: THREE.Object3D
  contact: {
    ni: number[]
    ri: number[]
    rj: number[]
    impactVelocity: number
  }
  collisionFilters: {
    bodyFilterGroup: number
    bodyFilterMask: number
    targetFilterGroup: number
    targetFilterMask: number
  }
}
```

Some things I care about when processing a collision:

- Contact point, in world space
- Normal of contact point
- Identifying data of the colliding object (i.e. a unique ID which I can reference back to the other entity, or some sort of arbitrary `userData` I can assign and retrieve on contact to reason about the other object)

Are these things provided? Is `ni` related to the normal, for example? Is `ri` or `rj` the contact point?

**Comment by a-type** [comment, marked as answer]
For anyone else looking, I think I found the answer in `cannon-es` source:

https://github.com/pmndrs/cannon-es/blob/master/src/equations/ContactEquation.ts#L16

```
  ri: Vec3 // World-oriented vector that goes from the center of bi to the contact point.
  rj: Vec3 // World-oriented vector that starts in body j position and goes to the contact point.
  ni: Vec3 // Contact normal, pointing out of body I.
```

Will update once I understand which body is `i` and which is `j` in relation to the contact event.

Update: maybe I'm reading this wrong, but it... looks like there's no way to really tell which body is `i` vs `j` for any particular contact event. The contact itself would normally have `bi` and `bj` fields, and I suppose you could compare them to the current body vs the colliding body to tell what `ri/rj/ni` mean. But the worker event doesn't provide those identities, so for one of the bodies the `ni` will be a normal from the target body, and for the other the `ni` will be for the colliding body, and I'm not sure there's a way with the current API to tell which is which.

**Reply by jak2030** [reply]
Thanks for figuring this out, helps a lot! Curious if anyone can verify this to be the case -- I am also hoping to distinguish between different objects in order to determine different collision behaviors.
